### PR TITLE
feat: implement virtionet

### PR DESCRIPTION
**Pull request by alexandrebrg**
One day this PR will be done :face_exhaling:

### CODE REVIEW

**Comment by sameo** [comment]
Let me know when/if you want me to review this PR, even if it's not fully ready yet.

**Comment by AlexandreBrg** [comment]
> Let me know when/if you want me to review this PR, even if it's not fully ready yet.

Sure, you can start to review until [lumper: add virtio queue consts](https://github.com/virt-do/lumper/pull/22/commits/d8c85f16ac86b939d5ba7c63fcb8129196c7cc26), thanks. The code is pretty much all taken from `vmm-reference`. Other commits cannot be reviewed for now, as they do not compile.

File: src/vmm/src/devices/mod.rs [hunk 1]
```diff
@@ -25,4 +59,317 @@ pub enum Error {
...
+    Endpoint(EvmgrError),
+}
+
+#[derive(Copy, Clone)]
+pub struct MmioConfig {
+    pub range: MmioRange,
+    // The interrupt assigned to the device.
+    // Global System Interrupts can be thought of as ACPI Plug and Play IRQ numbers.
+    // They are used to virtualize interrupts in tables and in ASL methods that perform resource allocation of interrupts.
+    // https://stackoverflow.com/questions/45206171/how-sci-system-control-interrupt-vector-is-defined
```
Comment (sameo) [inline src/vmm/src/devices/mod.rs:114]: I dont think this comment is relevant. `gsi`here is the hard-coded, legacy interrupt the virtio-mmio device will be bound to.
It's not related to ACPI's SCI.

Just keep `The interrupt assigned to the device (as passed to request_irq())`

File: src/vmm/src/devices/mod.rs [hunk 2]
```diff
@@ -25,4 +59,317 @@ pub enum Error {
...
+                self.mmio_cfg.gsi,
+                None,
+            )
+            .map_err(Error::Cmdline)?;
+
+        Ok(())
+    }
+
+    // Appends a string to the inner kernel cmdline.
+    pub fn insert_cmdline_str<T: AsRef<str>>(&mut self, t: T) -> Result<()> {
```
Comment (sameo) [inline src/vmm/src/devices/mod.rs:179]: I don't think you need this method at all.

File: src/vmm/src/devices/mod.rs [hunk 3]
```diff
@@ -25,4 +59,317 @@ pub enum Error {
...
+        self.kernel_cmdline
+            .insert_str(t.as_ref())
+            .map_err(Error::Cmdline)
+    }
+}
+
+// Holds configuration objects which are common to all current devices.
```
Comment (sameo) [inline src/vmm/src/devices/mod.rs:186]: They're common to all `virtio-mmio`devices.

File: src/vmm/src/devices/simple_handler.rs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+// A simple handler implementation for a RX/TX queue pair, which does not make assumptions about
+// the way queue notification is implemented.
+pub struct SimpleHandler<M: GuestAddressSpace, S: SignalUsedQueue> {
+    pub driver_notify: S,
+    pub rxq: Queue<M>,
+    pub rxbuf_current: usize,
+    pub rxbuf: [u8; MAX_BUFFER_SIZE],
+    pub txq: Queue<M>,
+    pub txbuf: [u8; MAX_BUFFER_SIZE],
+    pub tap: Tap,
```
Comment (sameo) [inline src/vmm/src/devices/simple_handler.rs:44]: Adding a `TAP`here makes it bound to networking, not just a generic, simple virtio queues handler. I'm expecting to see some sort of networking virtio device structure that would made of, at least, a `SimpleHandler`and a `Tap`.

**Comment by sameo** [comment]
#44 should supersede that one. Would you agree @AlexandreBrg ? If so, can we close that PR?

**Comment by alexandrebrg** [comment]
> #44 should supersede that one. Would you agree @AlexandreBrg ? If so, can we close that PR?

Sure @sameo, very glad to see that this heavy task has been integrated into the project!

**Pull request closed without merging**